// minecraft-server-util/dist/statusBedrock.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.statusBedrock = void 0;
const assert_1 = __importDefault(require("assert"));
const crypto_1 = __importDefault(require("crypto"));
const util_1 = require("util");
const Packet_1 = __importDefault(require("./structure/Packet"));
const UDPSocket_1 = __importDefault(require("./structure/UDPSocket"));
const resolveSRV_1 = __importDefault(require("./util/resolveSRV"));
const getTimeoutPromise_1 = __importDefault(require("./util/getTimeoutPromise"));
const parseDescription_1 = __importDefault(require("./util/parseDescription"));
const magic = [0x00, 0xFF, 0xFF, 0x00, 0xFE, 0xFE, 0xFE, 0xFE, 0xFD, 0xFD, 0xFD, 0xFD, 0x12, 0x34, 0x56, 0x78];
const ipAddressRegEx = /^\d{1,3}(\.\d{1,3}){3}$/;
function applyDefaultOptions(options) {
    // Apply the provided options on the default options
    return Object.assign({
        port: 19132,
        timeout: 1000 * 5,
        enableSRV: true,
        clientGUID: crypto_1.default.randomBytes(4).readUInt32BE()
    }, options);
}
/**
 * Retrieves the status of the Bedrock edition server
 * @param {string} host The host of the server
 * @param {BedrockStatusOptions} [options] The options to use when retrieving the status
 * @returns {Promise<BedrockStatusResponse>} The status information of the server
 * @async
 */
function statusBedrock(host, options) {
    var _a, _b;
    return __awaiter(this, void 0, void 0, function* () {
        // Applies the provided options on top of the default options
        const opts = applyDefaultOptions(options);
        // Assert that the arguments are the correct type and format
        assert_1.default(typeof host === 'string', `Expected 'host' to be a string, got ${typeof host}`);
        assert_1.default(host.length > 0, 'Expected \'host\' to have content, got an empty string');
        assert_1.default(typeof options === 'object' || typeof options === 'undefined', `Expected 'options' to be an object or undefined, got ${typeof options}`);
        assert_1.default(typeof opts === 'object', `Expected 'options' to be an object, got ${typeof opts}`);
        assert_1.default(typeof opts.port === 'number', `Expected 'options.port' to be a number, got ${typeof opts.port}`);
        assert_1.default(opts.port > 0, `Expected 'options.port' to be greater than 0, got ${opts.port}`);
        assert_1.default(opts.port < 65536, `Expected 'options.port' to be less than 65536, got ${opts.port}`);
        assert_1.default(Number.isInteger(opts.port), `Expected 'options.port' to be an integer, got ${opts.port}`);
        assert_1.default(typeof opts.timeout === 'number', `Expected 'options.timeout' to be a number, got ${typeof opts.timeout}`);
        assert_1.default(opts.timeout > 0, `Expected 'options.timeout' to be greater than 0, got ${opts.timeout}`);
        assert_1.default(typeof opts.enableSRV === 'boolean', `Expected 'options.enableSRV' to be a boolean, got ${typeof opts.enableSRV}`);
        assert_1.default(typeof opts.clientGUID === 'number', `Expected 'options.clientGUID' to be a number, got ${typeof opts.clientGUID}`);
        let srvRecord = null;
        // Automatically resolve from host (e.g. play.hypixel.net) into a connect-able address
        if (opts.enableSRV && !ipAddressRegEx.test(host)) {
            srvRecord = yield resolveSRV_1.default(host);
        }
        // Create a new UDP connection to the specified address
        const socket = new UDPSocket_1.default((_a = srvRecord === null || srvRecord === void 0 ? void 0 : srvRecord.host) !== null && _a !== void 0 ? _a : host, (_b = srvRecord === null || srvRecord === void 0 ? void 0 : srvRecord.port) !== null && _b !== void 0 ? _b : opts.port);
        // https://wiki.vg/Raknet_Protocol#Unconnected_Ping
        const pingPacket = new Packet_1.default();
        pingPacket.writeByte(0x01); // Packet ID
        pingPacket.writeLongBE(BigInt(Date.now())); // Time
        pingPacket.writeByte(...magic); // Magic
        pingPacket.writeLongBE(BigInt(opts.clientGUID)); // Client GUID
        socket.writePacket(pingPacket);
        // https://wiki.vg/Raknet_Protocol#Unconnected_Pong
        const pongPacket = yield socket.readPacket();
        // Begin reading the data out of the packet
        const packetType = pongPacket.readByte(); // Packet type
        // Packet was unexpected type, ignore the rest of the data in this packet
        if (packetType !== 0x1C)
            throw new Error('Server sent an invalid packet type');
        // Finish reading the data out of the packet
        pongPacket.readLongBE(); // Time
        const serverGUID = pongPacket.readLongBE(); // Server GUID
        pongPacket.readBytes(16); // Magic
        const serverIDBytes = pongPacket.readBytes(pongPacket.readUShortBE()); // Server ID
        // Property decode the server ID bytes into a response string
        const response = new util_1.TextDecoder().decode(Uint8Array.from(serverIDBytes));
        // Split the response string into multiple tokens, where each is a status item
        const splitResponse = response.split(';');
        // Grab each element out of the split response string, they may be missing which we'll fix later in the code
        const [edition, motdLine1, protocolVersion, version, onlinePlayers, maxPlayers, serverID, motdLine2, gameMode, gameModeID, portIPv4, portIPv6] = splitResponse;
        // Destroy the socket, it is no longer needed
        yield socket.destroy();
        return {
            host,
            port: opts.port,
            srvRecord,
            edition: edition && edition.length > 0 ? edition : null,
            serverGUID,
            motdLine1: motdLine1 && motdLine1.length > 0 ? parseDescription_1.default(motdLine1) : null,
            motdLine2: motdLine2 && motdLine1.length > 0 ? parseDescription_1.default(motdLine2) : null,
            version: version && version.length > 0 ? version : null,
            protocolVersion: isNaN(parseInt(protocolVersion)) ? null : parseInt(protocolVersion),
            maxPlayers: isNaN(parseInt(maxPlayers)) ? null : parseInt(maxPlayers),
            onlinePlayers: isNaN(parseInt(onlinePlayers)) ? null : parseInt(onlinePlayers),
            serverID: serverID && serverID.length > 0 ? serverID : null,
            gameMode: gameMode && gameMode.length > 0 ? gameMode : null,
            gameModeID: isNaN(parseInt(gameModeID)) ? null : parseInt(gameModeID),
            portIPv4: isNaN(parseInt(portIPv4)) ? null : parseInt(portIPv4),
            portIPv6: isNaN(parseInt(portIPv6)) ? null : parseInt(portIPv6)
        };
    });
}
/**
 * Retrieves the status of the Bedrock edition server
 * @param {string} host The host of the server
 * @param {StatusOptions} [options] The options to use when retrieving the status
 * @returns {Promise<BedrockStatusResponse>} The status information of the server
 * @async
 */
function statusWithTimeout(host, options) {
    var _a;
    return Promise.race([
        statusBedrock(host, options),
        getTimeoutPromise_1.default((_a = options === null || options === void 0 ? void 0 : options.timeout) !== null && _a !== void 0 ? _a : 1000 * 15, 'Failed to retrieve the status of the server within time')
    ]);
}
exports.statusBedrock = statusWithTimeout;
